webui/react/src/components/kit/CodeEditor.tsx

```diff
@@ -389,67 +404,73 @@ const CodeViewer: React.FC<Props> = ({
 
   // Set the selected node based on the active settings
   useEffect(() => {
-    if (!settings.filePath) return;
+    if (!fileViewerInfo.filePath) return;
 
-    if (settings.filePath && activeFile?.key !== settings.filePath) {
-      if (isConfig(settings.filePath)) {
-        handleSelectConfig(settings.filePath);
+    if (activeFile?.key !== fileViewerInfo.filePath) {
+      onSelectFile?.(fileViewerInfo.filePath);
+
+      if (isConfig(fileViewerInfo.filePath)) {
+        handleSelectConfig(fileViewerInfo.filePath);
       } else {
-        const path = settings.filePath.split('/');
+        const path = fileViewerInfo.filePath.split('/');
         const fileName = path[path.length - 1];
 
-        setIsFetchingFile(true);
-        fetchFile(settings.filePath, fileName);
-        switchTreeViewToEditor();
+        if (files?.length) {
+          setActiveFile({
+            key: path[path.length - 1],
+            text: fileViewerInfo.fileText,
+            title: path,
+          } as TreeNode);
+        } else {
+          setIsFetchingFile(true);
+          fetchFile(fileViewerInfo.filePath, fileName);
+        }
       }
     }
   }, [
     treeData,
-    settings.filePath,
+    files?.length,
+    fileViewerInfo,
     activeFile,
     fetchFile,
     handleSelectConfig,
-    switchTreeViewToEditor,
+    onSelectFile,
   ]);
 
   // Set the code renderer to ipynb if needed
   useEffect(() => {
     const hasActiveFile = activeFile?.text;
-    const isSameFile = activeFile?.key === settings.filePath;
-    const isIpybnFile = /\.ipynb$/i.test(settings.filePath);
+    const isSameFile = activeFile?.key === fileViewerInfo.filePath;
+    const isIpybnFile = /\.ipynb$/i.test(fileViewerInfo.filePath);
 
     if (hasActiveFile && isSameFile && isIpybnFile) {
       setEditorMode('ipynb');
     } else {
       setEditorMode('monaco');
```

Comment:
slightly out of scope, but can we change this to calculate on the fly?

Reply:
I can work on the other ones, but can you explain this question more? Having a separate component or something for the notebooks?

Reply:
`editorMode` is only set here as an effect of either `activeFile` or `fileViewerInfo` changing. Because `activeFile` and `fileViewerInfo` are both state themselves, changing either causes a possible double render when switching editor modes. Given that none of the constituent calculations are that expensive, we could just pull this code out, e.g:

```jsx
  const hasActiveFile = activeFile?.text;
  const isSameFile = activeFile?.key === fileViewerInfo.filePath;
  const isIpybnFile = /\.ipynb$/i.test(fileViewerInfo.filePath);

  const editorMode = (hasActiveFile && isSameFile && isIpybnFile) ? 'ipynb' : 'monaco'
```

If we're really adamant that we should only recalculate `editorMode` when `activeFile` or `fileViewerInfo` changes, we should use `useMemo` instead:
```jsx
  const editorMode = useEffect(() => {
    const hasActiveFile = activeFile?.text;
    const isSameFile = activeFile?.key === fileViewerInfo.filePath;
    const isIpybnFile = /\.ipynb$/i.test(fileViewerInfo.filePath);

    if (hasActiveFile && isSameFile && isIpybnFile) {
      return 'ipynb'
    } else {
      return 'monaco'
    }
  }, [fileViewerInfo.filePath, activeFile]);
```

Reply:
:+1: makes sense. I have a fix but am going to open this PR on the central repo so the Netlify is there